### PR TITLE
Htk spatial mapping tap

### PR DESCRIPTION
**Pull request by StephenHodgson**
https://github.com/Microsoft/HoloToolkit-Unity/issues/335 Implements @pinkwerks spatial tap shader from PR https://github.com/Microsoft/HoloToolkit-Unity/pull/353

spatialtap

Could use a bit of tweaking, but it's very similar to the tap pulse you observe in the shell.

Needs a couple improvements to the shader:

- [x] ~~Single & Stereo passes~~
- [ ] Reduce overdraw and increase overall performance
- [ ] Should be more like the occlusion shader, currently the spatial mapping tap shader uses black as the method of making the mesh invisible (really this only shows up on MR Video Captures).
- [ ] Leading and trailing edge falloff curves to customize the edge transitions to be sharper / smoother.
- [ ] Ability to fade at the end of its lifetime instead of just popping.

The SpatialMapping.prefab is set to now use this by default, and the Object Surface Observer component has been removed (It was mainly for testing in the emulator.  I believe prefabs should be production ready).

You can tweak all the setting through the spatial mapping manager:
image

I need someone to make sure all the test scenes and spatial understanding scenes work before we merge.

### CODE REVIEW

**Comment by HodgsonSDAS** [comment]
@thebanjomatic, could you check out the shader to make sure it works with single pass?

**Comment by thebanjomatic** [comment]
@HodgsonSDAS Will do. I already have a modified version of this shader from the old pull request. I'll let you know what changes need to be applied.

**Comment by HodgsonSDAS** [comment]
@thebanjomatic I've modified the shader a bit from the original.  You might want to try and pull this version.

**Comment by thebanjomatic** [comment]
@HodgsonSDAS I have submitted a pull request to the [HTK-SpatialMappingTap branch on your fork](https://github.com/HodgsonSDAS/HoloToolkit-Unity/pull/4) with the changes to enable stereo instancing.

**Comment by thebanjomatic** [comment]
Performance in a simple sceen with only this shader on the surface mesh and the `DefaultCursor` leaves a lot to be desired, it frequently drops to 30fps depending on where you are looking.  

It seems like this shader and the wireframe shader both suffer from lots of overdraw:  
image

I don't have 5.4 installed to check if the behavior is different there, but even the built-in unity wireframe shader suffers from the same problem in 5.5.0p2, this shader just does more.

File: Assets/HoloToolkit/SpatialMapping/Shaders/SpatialMappingTap.shader [hunk 1]
```diff
@@ -0,0 +1,126 @@
+Shader "Spatial Mapping/Spatial Mapping Tap"
```
Comment (thebanjomatic) [inline Assets/HoloToolkit/SpatialMapping/Shaders/SpatialMappingTap.shader]: This should probably be: "HoloToolkit/Spatial Mapping Tap"

**Comment by S-Hodgson-MMOARgames** [comment]
Not sure how to fix the overdraw, but I agree it is definitely a problem that not just the spatial mapping tap shader has, but all shaders that might try to visualize spatial mapping. It might be beneficial to try and figure out how to cull the geometry we're currently unable to see (or at least sections or chunks).

**Comment by thebanjomatic** [comment]
The best idea I was able to come up with (but haven't implemented) is to dynamically switch the shaders between Occlussion and SpatialMappingTap on a per-surface basis depending on whether or not the bounding box falls within the range of _Center + (_Radius +/- _PulseWidth). 

After playing around in the graphics debugger, I don't necessarily think there is an overdraw problem, its just amounts to running a relatively expensive fragment shader on every single pixel. I tried moving some of the color math (`_PulseColor * i.pulse` and `_WireframeColor * i.pulse`) out of the pixel shader and into the vertex shader, but that didn't help much because we are then doing extra work again on vertices that are ultimately occluded or off screen and won't be drawn.

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialMappingSource.cs [hunk 2]
```diff
@@ -3,11 +3,12 @@
 
 using System;
 using System.Collections.Generic;
+using HoloToolkit.Unity.InputModule;
 using UnityEngine;
 
 namespace HoloToolkit.Unity.SpatialMapping
 {
-    public class SpatialMappingSource : MonoBehaviour
+    public class SpatialMappingSource : MonoBehaviour, IInputClickHandler
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialMappingSource.cs]: I'll be sure to fix after new year's.

**Comment by HodgsonSDAS** [comment]
Thanks for looking into this.

**Comment by pinkwerks** [comment]
Hi all sorry I've been on vacation. Indeed running the fragment code is pretty much wasted on everything outside the pulse. You could try bailing out early with discard if your outside the pulse or getting smart about the spatial mapping chunks you apply the shader to. I'm still away from a computer for a couple weeks so I can't hack on it now. 

At the end of the day it's really a fill problem with hololens. Last I checked painting every pixel a solid color with spatial mapping wasn't performant.

**Comment by HodgsonSDAS** [comment]
No worries @pinkwerks, we're all away haha.
We'll still be here when you get back.

**Comment by thebanjomatic** [comment]
Good thinking @pinkwerks. I'm not sure discard is appropriate since then it won't write to the depth buffer, however, we could bail out early of the fragment shader with `return float4(0, 0, 0, 0);` if the pulse value of all three vertices is 0. That way there won't be any branch divergence per-triangle.

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialMappingSource.cs (hunk 2 above)
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialMappingSource.cs]: Done

**Comment by HodgsonSDAS** [comment]
> Similar short circuiting could happen if _Radius <= - _PulseWidth so that it acts (more or less) as a simple occlusion shader instead.

I can't wait to see this.

**Comment by thebanjomatic** [comment]
@HodgsonSDAS I'll take another look at this soon. I implemented the above changes and also switched to using `fixed` instead of `half` for a lot of the computation. I wanted to spend some time profiling on-device to verify that the performance improvement from reducing the precision warrants the change, but I can at least push it to my own fork this evening so that people can take a look.

**Comment by HodgsonSDAS** [comment]
Hey guys, check this out: https://github.com/keijiro/SonarFx

**Comment by pinkwerks** [comment]
re: sonarfx, I can't imagine that calling pow() twice per pixel is cheaper than doing the cubic pulse in the vertex shader.

If you don't need the wireframe effect, you could generate a shader permutation and get rid of the if statement - but I've never done that in unity. (multi compile?)

**Comment by HodgsonSDAS** [comment]
Re: SonarFx. Haha, I just thought it was cool and wanted to share.

**Pull request closed without merging**